bpaotu/bpaotu/frontend/src/pages/search_page/reducers/blast_search.ts

```diff
@@ -0,0 +1,193 @@
+
+const BLAST_SUBMISSION_POLL_FREQUENCY_MS = 5000
+
+export const {
+  handleBlastSequence,
+
+  runBlastStarted,
+  runBlastEnded,
+
+  blastSubmissionUpdateStarted,
+  blastSubmissionUpdateEnded,
+
+  clearBlastAlert
+} = createActions(
+  HANDLE_BLAST_SEQUENCE,
+
+  RUN_BLAST_STARTED,
+  RUN_BLAST_ENDED,
+
+  BLAST_SUBMISSION_UPDATE_STARTED,
+  BLAST_SUBMISSION_UPDATE_ENDED,
+
+  CLEAR_BLAST_ALERT
+)
+
+const blastInitialState = {
+  sequenceValue: '',
+  alerts: [],
+  isSubmitting: false,
+  submissions: []
+}
+
+export const runBlast = () => (dispatch, getState) => {
+  const state = getState()
+
+  dispatch(runBlastStarted())
+
+  const filters = describeSearch(state.searchPage.filters)
+  const searchString = state.searchPage.blastSearch.sequenceValue
+
+  executeBlast(searchString, filters)
+    .then(data => {
+      if (_get(data, 'data.errors', []).length > 0) {
+        dispatch(runBlastEnded(new ErrorList(data.data.errors)))
+        return
+      }
+      dispatch(runBlastEnded(data))
+      dispatch(autoUpdateBlastSubmission())
+    })
+    .catch(error => {
+      dispatch(runBlastEnded(new ErrorList('Unhandled server-side error!')))
+    })
+}
+export const autoUpdateBlastSubmission = () => (dispatch, getState) => {
+  const getLastSubmission: (() => BlastSubmission) = () => last(getState().searchPage.blastSearch.submissions)
+  const lastSubmission = getLastSubmission()
+  getBlastSubmission(lastSubmission.submissionId)
+    .then(data => {
+      dispatch(blastSubmissionUpdateEnded(data))
+      const newLastSubmission = getLastSubmission()
+      if (!newLastSubmission.finished) {
+        setTimeout(() => dispatch(autoUpdateBlastSubmission()), BLAST_SUBMISSION_POLL_FREQUENCY_MS)
+      }
+    })
+    .catch(error => {
+      dispatch(blastSubmissionUpdateEnded(new Error('Unhandled server-side error!')))
+    })
+}
+function alert(text, color = 'primary') {
+  return { color, text }
+}
+
+const BLAST_ALERT_IN_PROGRESS = alert('BLAST search is in Progress ...')
+const BLAST_ALERT_ERROR = alert('An error occured while running BLAST.', 'danger')
+
+export default handleActions(
+  {
+    [handleBlastSequence as any]: (state, action: any) => ({
+      ...state,
+      sequenceValue: join(filter(upperCase(action.payload), ch => includes('GATC', ch)), '')
+    }),
+    [runBlastStarted as any]: (state, action: any) => ({
+      ...state,
+      isSubmitting: true
+    }),
+    [runBlastEnded as any]: {
+      next: (state, action: any) => {
+        const lastSubmission: BlastSubmission = {
+          submissionId: action.payload.data.submission_id,
+          finished: false,
+          succeeded: false
+        }
+        const alerts = [BLAST_ALERT_IN_PROGRESS]
+
+        return {
+          ...state,
+          sequenceValue: state.sequenceValue,
+          alerts,
+          submissions: [...state.submissions, lastSubmission]
+        }
+      },
+      throw: (state, action: any) => ({
```

Comment:
You probably want to display some error message here?

Reply:
error alert added